packages/graph-explorer/src/core/StateProvider/localDb.ts

```diff
@@ -0,0 +1,223 @@
+  const backup = await createBackupData(localDb);
+  const serializedExportData = serializeData(backup);
+  logger.debug("Backup serialized for file", serializedExportData);
+  const fileData = toJsonFileData(serializedExportData);
+  logger.debug(`Converted data to JSON file blob`);
+  const fileName = "graph-explorer-config.json";
+  saveAs(fileData, fileName);
+  logger.debug(`Saved config data as ${fileName}`);
+}
+
+/** Reads the given blob, verifies it is a backup file, restores the data localDb. */
+export async function readBackupDataFromFile(blob: Blob) {
+  const jsonData = await fromFileToJson(blob);
+  logger.debug("Deserialized JSON data from file blob", jsonData);
+  const deserializedData = deserializeData(jsonData);
+  logger.debug(
+    "Deserialized JSON data from JSON serialization",
+    deserializedData
+  );
+  const result = await SerializedBackupSchema.parseAsync(deserializedData);
+  logger.debug("Parsed backup result", result);
+  return result;
+}
+
+/** Applies the given backup data to the localDb, overwriting all contents. */
+export async function restoreBackup(
+  backup: SerializedBackup,
+  localDb: LocalDb
+) {
+  logger.debug("Restoring backup to local db", backup);
+  await removeRestoredEntries(localDb);
+  const entries = applyRestorePrefixToEntries(backup.data);
+  await saveEntriesToDb(entries, localDb);
+  await removeExistingEntries(localDb);
+  await removePrefixFromRestoredEntries(localDb);
+
+  const appliedBackupState = await createBackupData(localDb);
+  logger.debug("Backup restored to local db", appliedBackupState);
+}
+
+export interface LocalDb {
+  keys(): Promise<string[]>;
+  clear(): Promise<void>;
+  getItem<T>(key: string): Promise<T | null>;
+  setItem<T>(key: string, value: T): Promise<T>;
+  removeItem(key: string): Promise<void>;
+}
+
+const SerializedBackupSchema = z.object({
+  backupSource: z.string(),
+  backupSourceVersion: z.string(),
+  backupVersion: z.string(),
+  backupTimestamp: z.coerce.date(),
+  data: z.record(z.string(), z.unknown()),
+});
+
+export type SerializedBackup = z.infer<typeof SerializedBackupSchema>;
+
+export async function createBackupData(localDb: LocalDb) {
+  await removeRestoredEntries(localDb);
+  const keys = await localDb.keys();
+  let backup = {};
+
+  logger.debug(`Exporting config data from localForage`, keys);
+
+  for (const key of keys) {
+    const value = await localDb.getItem(key);
+    logger.debug(`Adding ${key} to backup`, value);
+
+    backup = {
+      ...backup,
+      [`${key}`]: value,
+    };
+  }
+
+  const result: SerializedBackup = {
+    backupSource: "Graph Explorer",
+    backupSourceVersion: __GRAPH_EXP_VERSION__,
+    backupVersion: "1.0",
+    backupTimestamp: new Date(),
+    data: backup,
+  };
+
+  logger.debug(`Backup created`, result);
+
+  return result;
+}
+
+export async function removeRestoredEntries(localDb: LocalDb) {
+  const keys = (await localDb.keys()).filter(hasRestoredPrefix);
```

Comment:
What do the "normal" db keys look like here? Is it data from the graph that could theoretically begin with `RESTORED-ENTRY-`?